### PR TITLE
BarPlugin PerformanceWidget

### PR DESCRIPTION
**Pull request by denjukebox**
This PR implements 3 performance counter based widgets to display statistics about the system in the bar.

- Extendable PerformanceCounter base class
- CPU Widget
- Memory Widget
- Network Widget

The original idea was suggested by @adolfocorreia
https://github.com/workspacer/workspacer/issues/265

### CODE REVIEW

File: src/workspacer.Bar/Widgets/PerformanceWidget.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            var interfaces = category.GetInstanceNames();
+            if (!interfaces.Contains(interfaceName))
+            {
+                _interfaceName = interfaces.First();
+            }
+        }
+
+        protected override string GetText(float value)
+        {
+            var amount = value;
```
Comment (dalyIsaac) [inline src/workspacer.Bar/Widgets/PerformanceWidget.cs:136]: I don't think the `amount` variable is necessary.

File: src/workspacer.Bar/Widgets/PerformanceWidget.cs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            return Parts(Part($"{GetText(_counter.NextValue())}{icon}", null, null, ClickAction));
+        }
+
+        private string GetIconOrEmpty()
+        {
+            var icon = string.Empty;
+
+            try
+            {
+                icon = char.ConvertFromUtf32(int.Parse(Icon, System.Globalization.NumberStyles.HexNumber));
```
Comment (dalyIsaac) [inline src/workspacer.Bar/Widgets/PerformanceWidget.cs:66]: I think it'd be nice if we could allow custom text and emojis.

File: src/workspacer.Bar/Widgets/PerformanceWidget.cs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        }
+    }
+
+    public class NetworkPerformanceWidget : PerformanceWidgetBase
+    {
+        protected override string CounterCategory => "Network Interface";
+        protected override string CounterName => "Bytes Total/sec";
+        protected override string CounterInstance => _interfaceName;
+        protected override string Icon => "1F5A7";
+        protected override Action ClickAction => () => Process.Start("explorer.exe", @"shell:::{26EE0668-A00A-44D7-9371-BEB064C98683}\3");
```
Comment (dalyIsaac) [inline src/workspacer.Bar/Widgets/PerformanceWidget.cs:119]: Can you add a comment describing what this points to?

**Review by dalyIsaac — CHANGES_REQUESTED** [review]
This looks awesome!

Can you also provide some documentation in `/docs` for potential usage?

**Comment by adolfocorreia** [comment]
Great work on this new PerformanceWidget, @denjukebox!

I was not able to run/test your code, but if I may make a comment on this PR, I would suggest against calling `_counter.NextValue()` in the `GetParts()` function and poll the counter only in the timer callback instead. That was an issue I had in my implementation. The problem was that the CPU/memory widget might be redrawn because of other events (e.g. a TimeWidget update), and a new counter poll might be run before the next timer event. You could try to reproduce this behavior by setting the PerformanceWidget with a long update interval (e.g. 1 minute) and a TimeWidget with a short update interval (e.g. 1 second).

**Comment by dalyIsaac** [comment]
Thanks for adding the documentation. I think it'd be good to update based on @adolfocorreia's feedback.

**Comment by denjukebox** [comment]
@adolfocorreia Thanks for noticing the update issue! I'd seen the numbers updating a bit quickly put didn't put things together.

**Comment by adolfocorreia** [comment]
Glad I could help!

But I think it is still necessary to call `Context.MarkDirty()` in the timer callback, isn't it? Otherwise the PerformanceWidget might not be updated on the screen if there aren't other redraw events.

**Comment by dalyIsaac** [comment]
> Glad I could help!
> 
> But I think it is still necessary to call `Context.MarkDirty()` in the timer callback, isn't it? Otherwise the PerformanceWidget might not be updated on the screen if there aren't other redraw events.

@denjukebox I think this might be the case.

**Comment by denjukebox** [comment]
Yup totally missed that. I added it in again with an overridable update function for the timer.

**Comment by dalyIsaac** [comment]
For `NetworkPerformanceWidget`, I think it'd be worth adding that to select the appropriate instance, you can get the name from Device Manager:

image

would be instantiated as:

```csharp
new NetworkPerformanceWidget("Intel[R] Ethernet Connection [7] I219-V"),
```

**Comment by xd999e** [comment]
Any plans to merge this feature into main?

**Comment by denjukebox** [comment]
Apparently this was still blocked by @dalyIsaac his request for documentation. I've done this in the mean while but I guess this has to be checked off still? Been away from this project for a bit now.

**Review by dalyIsaac — APPROVED** [review]

**Comment by denjukebox** [comment]
@dalyIsaac  Thanks for the merge an grammar fixes!

**Comment by pranavpa8788** [comment]
Whenever I use any of the performance widgets I get the The type or namespace name 'NetworkPerformanceWidget' could not be found (are you missing a using directive or an assembly reference?). 

I am using the latest verison of workspacer. Do I have to manually compile the dll file?

**Pull request merged**